### ISSUE TITLE
audio recording and encoding

### ISSUE THREAD

**Issue by mkassner** [mikeboers/PyAV#182]
I m trying to implement microphone recording. 

this works:

``` python

import av
#device as input
container = av.open('none:0',format="avfoundation")
audio_stream = None
for i, stream in enumerate(container.streams):
    if stream.type == b'audio':
        audio_stream = stream
        break
if not audio_stream:
    exit()

#file container for output:
out_container = av.open('test.wav','w')
# out_stream = out_container.add_stream(codec_name = 'mp3',rate=44100)
out_stream = out_container.add_stream(template=audio_stream)
for i,packet in enumerate(container.demux(audio_stream)):
    print float(packet.pts*packet.stream.time_base)
    out_container.mux(packet)
    if i >500:
        break
out_container.close()
```

however I need the pts in the media for later so I m looking at mp4 with aac but I cant get it to work:

``` python

import av
import logging
logging.basicConfig(level=logging.DEBUG)

#device as input
container = av.open('none:0',format="avfoundation")
audio_stream = None
for i, stream in enumerate(container.streams):
    if stream.type == b'audio':
        audio_stream = stream
        break
if not audio_stream:
    exit()

#file container for output:
out_container = av.open('test.mp4','w')
out_stream = out_container.add_stream(codec_name = 'aac',rate=44100)
for i,packet in enumerate(container.demux(audio_stream)):
    for frame in packet.decode():
        print (float(frame.pts)/frame.time_base['den'])
        out_pack = out_stream.encode(frame)
        print out_pack
        if out_pack:
            print out_pack.pts 
            out_container.mux(packet)
    #exit and flushing
    if i >500:
        while True:
            out_pack = out_stream.encode(None)
            if out_pack:
                print out_pack.pts
                out_container.mux(packet)
            else:
                break
        break
out_container.close()
```

output: 

```
139869.792653
None
139869.804263
<av.Packet of #0, dts=None, pts=None at 0x108c48218>
None
139869.815873
<av.Packet of #0, dts=None, pts=None at 0x108c482b0>
None
139869.827483
<av.Packet of #0, dts=None, pts=None at 0x108c48348>
None
139869.839093
WARNING:libav.aac:Trying to remove 1024 samples, but the queue is empty
<av.Packet of #0, dts=None, pts=None at 0x108c48218>
None
139869.850703
None
139869.862313
None
...
```

The file does not play. 

Any ideas what I m doing wrong?

**Comment by mikeboers** [comment]
Sorry for the delays.

I'm at least super happy the first example is working for you (as I've been elbow deep in really broken PyAV for a while, and I guess my expectations are low), and the second does do the same for me. Hopefully I find a bit more time to dig into this in the coming days.

Aside: I need to document the `container.streams.get(audio=0)` stuff for stream selection. :wink:

**Comment by mikeboers** [comment]
I think there might be something wrong with the FIFOs due to the:

```
WARNING:libav.aac:Trying to remove 1024 samples, but the queue is empty
```

If the FIFO gets commented out of `av/audio.steam.pyx`, then it at least keeps having full packets. Still doesn't play though.

I really need to go become unconscious though. Hopefully more soon.

**Comment by mkassner** [comment]
Thats for the reply and again for your continued work on PyAV! I think A/V work is complex and messy and PyAV is the best option for me so far.

**Comment by mikeboers** [comment]
Even if I get past the FIFO stuff, am manually calculating known-correct PTS/DTSes, then I still get a ton of crap when I try to play it:

```
[aac @ 0x7fa0db80b000] Prediction is not allowed in AAC-LC.
[aac @ 0x7fa0db80b000] channel element 3.11 is not allocated
[aac @ 0x7fa0db80b000] Reserved bit set.
[aac @ 0x7fa0db80b000] Number of scalefactor bands in group (63) exceeds limit (49).
[aac @ 0x7fa0db80b000] Reserved bit set.
[aac @ 0x7fa0db80b000] Number of scalefactor bands in group (59) exceeds limit (49).
[aac @ 0x7fa0db80b000] Prediction is not allowed in AAC-LC.
[aac @ 0x7fa0db80b000] Reserved bit set.
[aac @ 0x7fa0db80b000] Prediction is not allowed in AAC-LC.
[aac @ 0x7fa0db80b000] channel element 2.0 is not allocated
[aac @ 0x7fa0db80b000] Sample rate index in program config element does not match the sample rate index configured by the container.
```

etc..

It seems to be completely related to the "aac" codec. It isn't happy about something...

**Comment by mkassner** [comment]
What about trying with mp3?

**Comment by mikeboers** [comment]
I found the problem... You won't like it.

The part in your second example that says `out_container.mux(packet)` should be `out_container.mux(out_pack)`; we've been muxing the original packets!

😆 

Both mp3 and aac work for me now, even with all the warnings. I just pushed 014cdb2 which has my attempts at "fixing" the fifo and PTS stuff. [See here](https://github.com/mikeboers/PyAV/blob/audioencode/rec-mp4.py).
